### PR TITLE
Add helper for combining ParameterizedData

### PR DESCRIPTION
**Pull request by jodastephen**
Combining instances is tricky, this helper makes it reliable
Also, fix incorrect used of Preconditions

### CODE REVIEW

File: modules/market/src/main/java/com/opengamma/strata/market/param/ParameterizedDataCombiner.java [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+
+  //------------------------------------------------------------------------- 
+  // creates an instance
+  private ParameterizedDataCombiner(ParameterizedData[] underlyings) {
+    ArgChecker.notEmpty(underlyings, "underlyings");
+    int size = underlyings.length;
+    this.underlyings = underlyings;
+    int[] lookup = new int[size];
+    for (int i = 1; i < size; i++) {
+      lookup[i] = lookup[i - 1] + underlyings[i - 1].getParameterCount();
```
Comment (jacbop) [inline modules/market/src/main/java/com/opengamma/strata/market/param/ParameterizedDataCombiner.java]: should we validate that none of the underlyings are null upon construction?

File: modules/market/src/main/java/com/opengamma/strata/market/param/ParameterizedDataCombiner.java [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+   */
+  public static ParameterizedDataCombiner of(ParameterizedData... instances) {
+    return new ParameterizedDataCombiner(instances);
+  }
+
+  //------------------------------------------------------------------------- 
+  // creates an instance
+  private ParameterizedDataCombiner(ParameterizedData[] underlyings) {
+    ArgChecker.notEmpty(underlyings, "underlyings");
+    int size = underlyings.length;
```
Comment (jacbop) [inline modules/market/src/main/java/com/opengamma/strata/market/param/ParameterizedDataCombiner.java]: Validate that size > 0, or is it okay to have an empty combiner?

File: modules/market/src/test/java/com/opengamma/strata/market/param/ParameterizedDataCombinerTest.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    assertEquals(test.getParameter(1), 2d);
+    assertEquals(test.getParameter(2), 3d);
+    assertEquals(test.getParameter(3), 4d);
+    assertEquals(test.getParameter(4), 5d);
+    assertEquals(test.getParameterMetadata(0), ParameterMetadata.empty());
+    assertThrows(() -> test.getParameter(-1), IndexOutOfBoundsException.class);
+    assertThrows(() -> test.getParameter(5), IndexOutOfBoundsException.class);
+    assertThrows(() -> test.getParameterMetadata(-1), IndexOutOfBoundsException.class);
+    assertThrows(() -> test.getParameterMetadata(5), IndexOutOfBoundsException.class);
+    assertThrowsIllegalArg(() -> ParameterizedDataCombiner.of());
```
Comment (jacbop) [inline modules/market/src/test/java/com/opengamma/strata/market/param/ParameterizedDataCombinerTest.java]: I see the test with empty (size = 0) underlyings, but can we add tests with one or more null underlyings. I assume that should also throw an illegal arg exception.

File: modules/market/src/main/java/com/opengamma/strata/market/param/ParameterizedDataCombiner.java (hunk 1 above)
Comment (jodastephen) [inline modules/market/src/main/java/com/opengamma/strata/market/param/ParameterizedDataCombiner.java]: The approach in Strata is not to validate for null, except when creating state. Strata also allows an IAE or NPE when validating state. So, since this code will not allow nulls into the state (because an NPE will be thrown in the constructor) we treat it as fine.

The rationale is that we have almost eliminated null in Strata, thus validating against it is, for the most part, pointless.

File: modules/market/src/test/java/com/opengamma/strata/market/param/ParameterizedDataCombinerTest.java (hunk 3 above)
Comment (jodastephen) [inline modules/market/src/test/java/com/opengamma/strata/market/param/ParameterizedDataCombinerTest.java]: As above, we don't tend to write tests for null.

File: modules/market/src/main/java/com/opengamma/strata/market/param/ParameterizedDataCombiner.java (hunk 1 above)
Comment (jacbop) [inline modules/market/src/main/java/com/opengamma/strata/market/param/ParameterizedDataCombiner.java]: Can we add something along these lines to the code standards in the wiki? I did a search there before making these comments, but your explanation makes sense. I'll merge it now.
